src/mito2/src/sst/parquet/page_reader.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Parquet page reader.
+
+use std::collections::VecDeque;
+
+use parquet::column::page::{Page, PageMetadata, PageReader};
+use parquet::errors::Result;
+
+/// A reader that reads from cached pages.
+pub(crate) struct CachedPageReader {
+    /// Cached pages.
+    pages: VecDeque<Page>,
+}
+
+impl CachedPageReader {
+    /// Returns a new reader from existing pages.
+    pub(crate) fn new(pages: &[Page]) -> Self {
+        Self {
+            pages: pages.iter().cloned().collect(),
+        }
+    }
+}
+
+impl PageReader for CachedPageReader {
+    fn get_next_page(&mut self) -> Result<Option<Page>> {
+        Ok(self.pages.pop_front())
+    }
+
+    fn peek_next_page(&mut self) -> Result<Option<PageMetadata>> {
+        Ok(self.pages.front().map(page_to_page_meta))
+    }
+
+    fn skip_next_page(&mut self) -> Result<()> {
+        // When the `SerializedPageReader` is in `SerializedPageReaderState::Pages` state, it never pops
+        // the dictionary page. So it always return the dictionary page as the first page. See:
+        // https://github.com/apache/arrow-rs/blob/1d6feeacebb8d0d659d493b783ba381940973745/parquet/src/file/serialized_reader.rs#L766-L770
+        // But the `GenericColumnReader` will read the dictionary page before skipping records so it won't skip dictionary page.
+        // So we don't need to handle the dictionary page specifically in this method.
+        // https://github.com/apache/arrow-rs/blob/65f7be856099d389b0d0eafa9be47fad25215ee6/parquet/src/column/reader.rs#L322-L331
+        self.pages.pop_front();
+        Ok(())
+    }
+}
+
+impl Iterator for CachedPageReader {
```

Comment:
We know the size, maybe `ExactSizeIterator` is better.

https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html

Reply:
Not very necessary, parquet doesn't use it. Its `SerializedPageReader` also doesn't implement this trait.